spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,265 @@
+Defined as
+
+```go
+type ConsensusState struct {
+  Base VerifierBase
+  Root CommitmentRoot
+}
+```
+where
+  * `Base` is a data used by `Consensus.Verifier` to verify `Header`s
+  * `Root` is the `CommitmentRoot`, used to prove internal state
+
+##### Header
+
+`Header` is a blockchain header which provides information to update `ConsensusState`,
+submitted to one blockchain to update the stored `ConsensusState`.
+Defined as
+
+```go
+type Header struct {
+  Proof HeaderProof
+  Base Maybe[VerifierBase]
+  Root CommitmentRoot
+}
+```
+where
+  * `Proof` is the commit proof used by `Consensus.Verifier` to be verified
+  * `Base` is the new verify, if needed to be updated
+  * `Root` is the new `CommitmentRoot` which will replace the existing one
+
+##### Verifier
+
+`Verifier` is a light client verifier proving `Header` depending on the `Commit`.
+It SHOULD prove far more efficiently than replaying `Consensus` logic
+for the given parent `Header` and the list of network messages, ideally in O(1) time.
+Defined as
+
+```go
+type Verifier func(ConsensusState, Header) (Error|ConsensusState)
+```
+
+The detailed specification of `Verifier` is defined in [VERIFIER.md](./VERIFIER.md)
+
+##### LightClient
+
+LightClient is defined as
+```go
+type LightClient struct {
+  Verifier Verifier
+  ConsensusState ConsensusState
+}
+```
+where
+  * `ConsensusState` is the root of trust providing the `VerifierBase`
+  * `Verifier` is the lightclient verification logic
+
+The exact type of each fields are depending on the type of the actual consensus logic.
+
+#### Subprotocols
+
+The chains MUST implement function `register` and `update`, as they form the `handleDatagram`.
+Calling both functions MAY be permissionless.
+
+##### Preliminaries
+
+`newID` is a function which generates a new `Identifier` for a `LightClient`, which MAY depending
+on the `Header`. The behaviour of `newID` is implementation specific. Possible implementations are:
+
+* Random bytestring
+* Hash of the `Header`
+* Incrementing integer index, bigendian encoded
+
+`newID` MUST NOT return an `Identifier` which has already been generated, so it can be stateful to check
+the `Identifier`
+
+`storekey` takes an `Identifier` and returns a `KVStore` compatible `Key`.
+
+```coffee
+newID = (header) -> # impl specific
+storekey = (id) -> # impl specific
+```
+
+##### Register
+
+Registering new `LightClient` is done simply by submitting it to the `register` function,
+as the chain automatically generates the `Identifier` for the `LightClient`.
+
+```coffee
+register = (lightclient) ->
+  id = newID()
+  set(storekey(id), register)
+  id
+}
+```
+
+##### Update
+
+Updating `LightClient` is done by submitting a new `Header`. The `Identifier` is used to point the
+stored `LightClient` that the logic will update. When the new `Header` is verifiable with
+the stored `LightClient`'s `Verifier` and `ConsensusState`, then it SHOULD update the
```

Comment:
```suggestion
the stored `LightClient`'s `Verifier` and `ConsensusState`, then it **should** update the
```

Reply:
I think we settled with this https://tools.ietf.org/html/rfc2119 as a standard on how to use should, must etc.

Reply:
Oh ok nvm then, but should this be indicated somewhere? @cwgoes @mossid

Reply:
It is indicated in ICS 1